### PR TITLE
kyc: add quiz

### PR DESCRIPTION
**Pull request by ice-dionysos**

### CODE REVIEW

File: kyc/quiz/DDL.sql [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    ended_successfully boolean    not null default false,
+    questions          bigint[]   not null,
+    answers            smallint[] not null,
+    user_id            text       primary key references users (id) ON DELETE CASCADE,
+    language           text       not null,
+    FOREIGN KEY (language, quiz_id) REFERENCES quizzes (language, id) ON DELETE CASCADE
+) WITH (FILLFACTOR = 70);
+
+
+insert into quizzes (created_at, language) values (now(), 'en') on conflict do nothing;
```
Comment (ice-ares) [inline kyc/quiz/DDL.sql]: this doesn't belong here. Cuz its seed logic for tests

File: kyc/quiz/contract.go [hunk 2]
```diff
@@ -14,6 +22,20 @@ const (
 )
 
 type (
+	UserID      = users.UserID
+	UserProfile = users.UserProfile
+
+	Manager interface {
```
Comment (ice-ares) [inline kyc/quiz/contract.go]: Repository

File: kyc/quiz/contract.go [hunk 3]
```diff
@@ -14,6 +22,20 @@ const (
...
+		io.Closer
+
+		Start(ctx context.Context, userID UserID, lang string, total int) (Quiz, error)
```
Comment (ice-ares) [inline kyc/quiz/contract.go]: StartQuizSession.

Whats that total for? If its those 21 questions, thats a config thing, not input

Comment (ice-ares) [reply kyc/quiz/contract.go]: *Quiz

File: kyc/quiz/contract.go [hunk 4]
```diff
@@ -14,6 +22,20 @@ const (
...
+		Continue(ctx context.Context, userID UserID, question uint, answer uint8) (Quiz, error)
```
Comment (ice-ares) [inline kyc/quiz/contract.go]: ContinueQuizSession
*Quiz

Where is the session identifier? U need to always match everything under the same identifier. I think its quiz_id

File: kyc/quiz/contract.go [hunk 5]
```diff
@@ -27,8 +49,58 @@ type (
 		IncorrectAnswers uint8      `json:"incorrectAnswers" example:"2"`
 	}
 	Question struct {
-		Text    string   `json:"text" example:"Какая температура на улице?"`
-		Options []string `json:"options" example:"+21,-2,+33,0"`
-		Number  uint8    `json:"number" example:"1"`
+		Text    string   `json:"text" example:"Какая температура на улице?" db:"question"`
+		Options []string `json:"options" example:"+21,-2,+33,0" db:"options"`
+		Number  uint     `json:"number" example:"1" db:"id"`
```
Comment (ice-ares) [inline kyc/quiz/contract.go]: nope. That number has nothing to do with the id.  Id is like a numeric uuid. Number is the Nth question in the quiz

File: kyc/quiz/contract.go [hunk 6]
```diff
@@ -27,8 +49,58 @@ type (
...
+	ErrSessionFinished         = errors.New("session closed")
+	ErrSessionExpired          = errors.New("session expired")
+	ErrUnknownQuestionNumber   = errors.New("unknown question number")
+	ErrUnknownSession          = errors.New("unknown session and/or user")
+)
+
+const (
+	applicationYamlKey = "kyc/quiz"
+
+	defaultMaxSessionDurationSeconds   = 60 * 60 * 1  // 1 hour.
```
Comment (ice-ares) [inline kyc/quiz/contract.go]: nope. everything is cfg based

File: kyc/quiz/contract.go [hunk 7]
```diff
@@ -27,8 +49,58 @@ type (
...
+	defaultQuestionsCacheExpireSeconds = 60 * 60 * 24 // 1 day.
+)
+
+var (
+	//go:embed DDL.sql
+	ddl string
+
+	_ UserReader = users.ReadRepository(nil)
```
Comment (ice-ares) [inline kyc/quiz/contract.go]: don't need this

File: kyc/quiz/contract.go [hunk 8]
```diff
@@ -27,8 +49,58 @@ type (
...
+)
+
+type (
+	quizDetails struct {
+		Questions []Question
+		ID        int
+	}
+
+	managerImpl struct {
```
Comment (ice-ares) [inline kyc/quiz/contract.go]: repository

File: kyc/quiz/contract.go [hunk 9]
```diff
@@ -27,8 +49,58 @@ type (
...
+		DB       *storage.DB
+		Shutdown func() error
+		Users    UserReader
+		Cache    struct {
```
Comment (ice-ares) [inline kyc/quiz/contract.go]: lets not add any caching for now. Get everything from DB all the time.

File: kyc/quiz/quiz_test.go [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+	require.Equal(t, uint8(2), session.Progress.MaxQuestions)
+	require.NotEmpty(t, session.Progress.NextQuestion.Text)
+
+	ans := helperSolveQuestion(t, session.Progress.NextQuestion.Text)
+	t.Logf("q: %v, ans: %d", session.Progress.NextQuestion.Text, ans)
+	session, err = m.Continue(ctx, "bogus", session.Progress.NextQuestion.Number, ans)
+	require.NoError(t, err)
+	require.NotNil(t, session)
+	require.Empty(t, session.Result)
+	require.NotNil(t, session.Progress)
```
Comment (ice-ares) [inline kyc/quiz/quiz_test.go:261]: not nil-ing everything adds little value. It's more noice than anything. Quality coverage is with EqualValues

File: kyc/quiz/quiz.go [hunk 11]
```diff
@@ -1,3 +1,448 @@
...
+	"github.com/ice-blockchain/wintr/time"
+)
+
+func loadConfig() config {
+	var cfg config
+
+	appcfg.MustLoadFromKey(applicationYamlKey, &cfg)
+
+	if cfg.MaxSessionDurationSeconds == 0 {
+		cfg.MaxSessionDurationSeconds = defaultMaxSessionDurationSeconds
```
Comment (ice-ares) [inline kyc/quiz/quiz.go]: nope, we panic

File: kyc/quiz/quiz.go [hunk 12]
```diff
@@ -1,3 +1,448 @@
...
+	}
+
+	if total > len(details.Questions) {
+		return 0, nil, errors.Wrapf(ErrNotEnoughData, "%v: requested=%d but available=%d", lang, total, len(details.Questions))
+	}
+
+	questions := make([]Question, 0, total)
+	used := make(map[uint8]struct{}, total)
+	for len(questions) != total {
+		n := fastrand.Uint32n(uint32(len(details.Questions)))
```
Comment (ice-ares) [inline kyc/quiz/quiz.go]: nope. We shuffle the questions in db, before we create quizz_sessions entry for user

File: kyc/quiz/quiz.go [hunk 13]
```diff
@@ -1,3 +1,448 @@
...
+}
+
+func (m *managerImpl) CacheQuestions(ctx context.Context) error {
+	type QuestionLang struct {
+		Lang string `db:"language"`
+		Question
+		Parent int `db:"quiz_id"`
+	}
+	const stmt = `
+select q."id", q."quiz_id", q."language", q."question", q."options"
```
Comment (ice-ares) [inline kyc/quiz/quiz.go]: oh, I think I understand the issue. We need to change the ddl a bit cuz we have a pool of X questions for each language, that we shuffle then pick 21 of them an assign them to a user, for that session

Imma give u the updated ddl

File: kyc/quiz/quiz.go [hunk 14]
```diff
@@ -1,3 +1,448 @@
...
+		s = append(s, questions[i].Number)
+	}
+
+	return s
+}
+
+func (m *managerImpl) preCheckUser(ctx context.Context, userID UserID) error {
+	profile, err := m.Users.GetUserByID(ctx, userID)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) || errors.Is(err, storage.ErrRelationNotFound) {
```
Comment (ice-ares) [inline kyc/quiz/quiz.go]: storage.ErrRelationNotFound not gonna happen

File: kyc/quiz/quiz.go [hunk 15]
```diff
@@ -1,3 +1,448 @@
...
+			return errors.Wrap(ErrUnknownUser, userID)
+		}
+
+		return errors.Wrapf(err, "failed to get user by id: %v", userID)
+	}
+
+	if profile.KYCStepPassed == nil || *profile.KYCStepPassed != users.Social1KYCStep {
```
Comment (ice-ares) [inline kyc/quiz/quiz.go]: use quizStep -1 , so we can move things around if we need to

File: kyc/quiz/quiz.go [hunk 16]
```diff
@@ -1,3 +1,448 @@
...
+	return
+}
+
+func (m *managerImpl) Continue(ctx context.Context, userID UserID, question uint, answer uint8) (Quiz, error) { //nolint:funlen // It's expected.
+	// $1: user_id
+	// $2: question
+	// $3: answer
+	// $4: max_session_duration.
+	const pipeline = `
+	with session_data as (
```
Comment (ice-ares) [inline kyc/quiz/quiz.go]: pls move the logic in go code. This is very hard to follow and maintain

File: kyc/quiz/quiz.go [hunk 17]
```diff
@@ -1,3 +1,448 @@
...
+	indexes := questionsToSlice(questions)
+
+	// $1: user_id
+	// $2: quiz_id
+	// $3: questions
+	// $4: lang
+	// $5: max_session_duration
+	// $6: current time.
+	const stmt = `
+	with sesssion_status as (
```
Comment (ice-ares) [inline kyc/quiz/quiz.go]: pls move the logic in go code. This is very hard to follow and maintain

File: kyc/quiz/quiz.go (hunk 13 above)
Comment (ice-ares) [reply kyc/quiz/quiz.go]: I've updated the DDL in the task description.

File: kyc/quiz/contract.go (hunk 4 above)
Comment (ice-dionysos) [reply kyc/quiz/contract.go]: >Where is the session identifier? U need to always match everything under the same identifier. I think its quiz_id

I didn't get it. 
We have `user_id text primary key references users (id)` inside `quizz_sessions` table so we can always match a user to a session. 
Inside the session table, we have everything we need already (`quiz_id` is also present).

Comment (ice-ares) [reply kyc/quiz/contract.go]: I was thinking to avoid race conditions, but I don't think its possible anymore

Comment (ice-ares) [reply kyc/quiz/contract.go]: Think of some scenarios that involve starting quiz on 2 separate devices, and check if u can have race conditions there, in any combination of requests

Comment (ice-dionysos) [reply kyc/quiz/contract.go]: Once the first session is created, it's not possible to create a second one (until some conditions ...) because of that primary key (user_id). So I would say we are safe here.

Comment (ice-ares) [reply kyc/quiz/contract.go]: Nope. you create a session, finish it (successfully or not), then after 1 week, u can do it again if u failed it previously. So probably we need another table of failed attempts for historical purposes 🤔

Comment (ice-ares) [reply kyc/quiz/contract.go]: updated the DDL again
